Source/JavaScriptCore/runtime/RegExpObjectInlines.h

```diff
@@ -30,9 +30,47 @@
 #include "RegExpGlobalDataInlines.h"
 #include "RegExpMatchesArray.h"
 #include "RegExpObject.h"
+#include "RegExpPrototype.h"
 
 namespace JSC {
 
+ALWAYS_INLINE bool RegExpObject::isSymbolReplaceFastAndNonObservable()
+{
+    JSGlobalObject* globalObject = this->globalObject();
+    // If regExpPrimordialPropertiesWatchpointSet is valid, we know that RegExp.prototype.exec,
+    // RegExp.prototype.flags, all flag getters, and RegExp.prototype[@@replace] are unmodified.
+    if (!globalObject->regExpPrimordialPropertiesWatchpointSet().isStillValid())
+        return false;
+
+    // If stringSymbolReplaceWatchpointSet is valid, we know String.prototype and Object.prototype
+    // don't have Symbol.replace added.
+    if (!globalObject->stringSymbolReplaceWatchpointSet().isStillValid())
+        return false;
+
+    // Check that lastIndex is a number for fast path.
+    if (!getLastIndex().isNumber())
+        return false;
+
+    VM& vm = globalObject->vm();
+    Structure* structure = this->structure();
+    // Fast case: If the structure is the original RegExp structure, we know there are no own properties.
+    if (structure == globalObject->regExpStructure()) [[likely]]
+        return true;
+
+    if (structure->hasPolyProto())
+        return false;
+
+    if (structure->storedPrototype() != globalObject->regExpPrototype())
+        return false;
+
+    // Otherwise, check that the RegExpObject doesn't have its own Symbol.replace property using
+    // getDirectOffset which doesn't throw.
+    if (getDirectOffset(vm, vm.propertyNames->replaceSymbol) != invalidOffset)
```

Comment:
Is it sufficient to just check `@@replace`? Don't we need to check the flag getters here as well?

Reply:
Ah sorry you updated to use `hasCustomProperty` already